Docs: Move in section on QueryBuilder to How-To (#4080)
aiidateam/aiida-core: docs/source/howto/data.rst

.. _how-to:data:find:

Finding and querying for data
=============================

Once you have successfully completed a series of workflows for your project, or have imported a dataset you are interested in, you want to quickly find the data that is relevant for your analysis.

> The data in an AiiDA database is stored as a graph of connected entities, which can be easily *queried* with the ``QueryBuilder`` class.

Reviewer: Make sure to actually link the class to the API reference (here and everywhere else).

Before starting to write a query, it helps to:

*   know what you want to query for.
    In the language of databases, you need to tell the backend what to *project*.
    For example, you might be interested in the label of a calculation and the PK's of all its outputs.
*   know the relationships between entities you are interested in.
    Nodes of an AiiDA graph (vertices) are connected with links (edges).
    A node can for example be either the input or output of another node, but also an ancestor or a descendant.
*   know how you want to filter the results of your query.

Once you are clear about what you want and how you can get it, the QueryBuilder will build an SQL-query for you.

There are two ways of using the QueryBuilder:

#.  In the *appender* method, you construct your query step by step using the ``QueryBuilder.append()`` method.
#.  In the *queryhelp* approach, you construct a dictionary that defines your query and pass it to the ``QueryBuilder``.

Both APIs provide the same functionality - the appender method may be more suitable for interactive use, e.g., in the ``verdi shell``, whereas the queryhelp method can be useful in scripting.

> In this how-to we will focus on the basics of the appender method.

Reviewer: ```suggestion
In this section we will focus on the basics of the appender method.
```

For more advanced queries or more details on the queryhelp, see the :ref:`topics section on advanced querying <topics:database:advancedquery>`.

.. _how-to:data:find:select:

Selecting entities
------------------

Using the ``append()`` method of the ``QueryBuilder``, you can query for the entities you are interested in.
Suppose you want to query for calculation job nodes in your database:

.. code-block:: python

    from aiida.orm.querybuilder import QueryBuilder

>     qb = QueryBuilder()       # Instantiating instance. One instance -> one query

Author: I think that's fine, it's definitely good to be consistent. Could it be confusing for users that `builder` is used for both the *process* builder and *query* builder?

>     qb = QueryBuilder()       # Instantiating instance. One instance -> one query

Reviewer: I think it could.

>     qb = QueryBuilder()       # Instantiating instance. One instance -> one query

Reviewer 2: I think it will always be clear from the context, but like I said in another thread in response to Simon, maybe we should indeed agree on the preferred variable name for these. I just want to make sure we are using it consistently and do not have a million different names across the code and documentation.

>     qb = QueryBuilder()       # Instantiating instance. One instance -> one query

Author: I'll leave it as `qb` for now, until we decide on the preferred variable name.

    qb.append(CalcJobNode)    # Setting first vertex of path

> If you are interested in instances of different classes, you can also pass a tuple, list or set of classes.

Reviewer: ```suggestion
If you are interested in instances of different classes, you can also pass an iterable of classes.
```
That's the Python term and it will be clear that this includes lists from the example.

However, they have to be of the same ORM-type (e.g. all have to be subclasses of ``Node``):

.. code-block:: python

    qb = QueryBuilder()       # Instantiating instance. One instance -> one query
    qb.append([CalcJobNode, WorkChainNode]) # Setting first vertice of path, either WorkChainNode or Job.

.. note::

>     Remember that :ref:`processes<topics:processes:concepts:types>` have both a run-time ``Process`` that steers them and a ``Node`` that stores their metadata in the database.

Reviewer: 1. What exactly does *steers* mean in this context? (I of course have an idea, but we should be precise.)
2. A ``Node`` can store more than metadata, right?

Author: 1. Perhaps use "executes" instead or "steers"?
2. Correct! I've changed "metadata" to "data".

Reviewer: Re 1: I think the issue is more that the relationship between the mentioned entities is unclear. But I don't want to hold up the PR because of this. If we can't clarify that, let's just leave it *as-is*.

Reviewer 2: I agree that it is not an easy-to-grasp concept, which is why I explained it in detail in [this topic section](https://aiida.readthedocs.io/projects/aiida-core/en/docs-revamp/topics/processes/concepts.html#concepts). Maybe referring to that for details would be sufficient.

>     The QueryBuilder allows you to pass either the ``Node`` class (e.g. ``CalcJobNode``) or the ``Process`` class (e.g. ``CalcJob``), which will automatically apply the correct filters for the type of calculation.

Reviewer: I'm honestly not quite sure what this exactly means. Can we use either or? Will this lead to the same/different results? Should I prefer one over the other?

Author: If I understand correctly, you can indeed use either and it will produce the same result. I suppose ``CalcJobNode`` is better to use, as it is the actual entity you are querying for. If the user *does* use ``CalcJob``, it should not affect their result. Maybe @sphuber can confirm this?

Perhaps this is a detail that is best moved to the topics section. If we do leave it here, maybe the following note is more clear:

```
.. note::

    Remember that :ref:`processes<topics:processes:concepts:types>` have both a run-time ``Process`` that executes them and a ``Node`` that stores their data in the database.
    The QueryBuilder allows you to pass either the ``Node`` class (e.g. ``CalcJobNode``) or the ``Process`` class (e.g. ``CalcJob``), which will automatically select the right entity for the query.
    So, using either ``CalcJobNode`` or ``CalcJob`` will produce the same query results.
```

Reviewer 2: By definition the QueryBuilder can only be used to query for Nodes, or subclasses thereof (and of course the other entities Computer, User, etc.). However, `Process` or any of its subclasses, is not a subclass of `Node`. Therefore, before you could not query for `PwCalculation` in the QueryBuilder. What you had to do instead was:
```
QueryBuilder().append(CalcJobNode, filters={'attributes.process_type': 'aiida.calculations:quantumespresso.pw'})
```
The `process_type` as you can see is the fully qualified entry point name of `PwCalculation`. Since this is cumbersome, I implemented support for appending process classes. This would interpret `append(PwCalculation)` and expand it automatically into the node class and filters as I described.

.. _how-to:data:find:results:

Retrieving results
------------------

Once you have *appended* the entity you want to query for to the ``QueryBuilder``, the next question is how to get the results.
There are several ways to obtain data from a query:

.. code-block:: python

    qb = QueryBuilder()                 # Instantiating instance
    qb.append(CalcJobNode)              # Setting first vertice of path

    first_row = qb.first()              # Returns a list (!)
                                        # of the results of the first row

    all_results_d = qb.dict()           # Returns all results as
                                        # a list of dictionaries

    all_results_l = qb.all()            # Returns a list of lists

In case you are working with a large dataset, you can also return your query as a generator:

.. code-block:: python

    all_res_d_gen = qb.iterdict()       # Return a generator of dictionaries
                                        # of all results

>     all_res_l_gen = qb.iterall()        # Returns a generator of lists

Author: I think it would be good to add an example, yes. E.g.:

```
.. code-block:: python

    all_res_d_gen = qb.iterdict()       # Return a generator of dictionaries
                                        # of all results
    all_res_l_gen = qb.iterall()        # Returns a generator of lists

This will retrieve the data in batches, and you can start working with the data before the query has completely finished.
For example, you can iterate over the results of your query in a for loop:

.. code-block:: python

    for entry in builder.iterall():
        # do something with a single entry in the query result
```

This will retrieve the data in batches, and you can start working with the data before the query has completely finished.

.. _how-to:data:find:filters:

Filters
-------

Usually you do not want to query for *all* entities of a certain class, but rather *filter* the results based on certain properties.

> Suppose you do not want to all ``CalcJobNode`` data, but only the ones that are ``finished``:

Reviewer: ```suggestion
Suppose you do not want all ``CalcJobNode`` data, but only those that are ``finished``:
```

.. code-block:: python

    qb = QueryBuilder()                 # Initialize a QueryBuilder instance
    qb.append(
        CalcJobNode,                    # Append a CalcJobNode
        filters={                       # Specify the filters:
            'attributes.process_state': 'finished',  # the process is finished
        },
    )

You can apply multiple filters to one entity in a query.

> Say you are interested in all calculation jobs in your database that are ``finished`` **and** have the ``process_label`` ``ArithmeticAddCalculation``:

Author: That's a good point, but aren't all processes with a numerical `exit_status` finished? So in case you specify the `'attributes.exit_status': 0` filter, isn't the `'attributes.process_state': 'finished'` filter redundant?

Maybe we can use `'attributes.resources.num_machines': 1` as the second filter, or is this too complicated?

> Say you are interested in all calculation jobs in your database that are ``finished`` **and** have the ``process_label`` ``ArithmeticAddCalculation``:

Reviewer 2: > That's a good point, but aren't all processes with a numerical `exit_status` finished?

Yes, that is true. But I think the "hidden inefficiency" does not outweigh the advantage of the clarity of the example. I indeed think the resources alternative may be a bit too complicated.

            'attributes.process_label': 'ArithmeticAddCalculation'    # has process_label ArithmeticAddCalculation
        },
    )

In case you want to query for calculation jobs that satisfy one of these conditions, you can use the ``or`` operator:

.. code-block:: python

    qb = QueryBuilder()
    qb.append(
        CalcJobNode,
        filters={
            'or':[
                {'attributes.process_state': 'excepted'},
                {'attributes.process_label': 'ArithmeticAddCalculation'}
            ]
        },
    )

If we had written ``and`` instead of ``or`` in the example above, we would have performed the exact same query as the previous one, because ``and`` is the default behavior if you provide several filters as key-value pairs in a dictionary to the ``filters`` argument.
In case you want all calculation jobs with state ``finished`` or ``excepted``, you can also use the ``in`` operator:

.. code-block:: python

    qb = QueryBuilder()
    qb.append(
        CalcJobNode,
        filters={
            'attributes.process_state': {'in': ['finished', 'excepted']}
        },
    )

> You can also negate a filter by adding an exclamation mark in front of the operator.

Reviewer: ```suggestion
You can negate a filter by adding an exclamation mark in front of the operator.
```

So, to query for all calculation jobs that are not a 'finished' or 'excepted' state:

.. code-block:: python

    qb = QueryBuilder()
    qb.append(
        CalcJobNode,
        filters={
            'attributes.process_state': {'!in': ['finished', 'excepted']}
        },
    )

.. note::

    The above rule applies to all operators.
    For example, you can check non-equality with ``!==``, since this is the equality operator (``==``) with a negation prepended.

> A list of all implemented operators can be found in the ``QueryBuilder`` docstring.

Reviewer: ```suggestion
A complete list of all available operators is documented as part of the ``QueryBuilder`` class documentation.
```

.. _how-to:data:find:relationships:

Relationships
-------------

It is possible to query for data based on its relationship to another entity in the database.
Imagine you are not interested in the calculation jobs themselves, but in one of the outputs they create.
You can build upon your initial query for all ``CalcJobNode``'s in the database using the relationship of the output to the first step in the query:

.. code-block::

    qb = QueryBuilder()
    qb.append(CalcJobNode, tag='calcjob')
    qb.append(Int, with_incoming='calcjob')

> In the first ``append`` call, we query for all ``CalcJobNode``'s in the database, and *tag* this step with the *unique* identifier ``'calcjob'``.

Author: I've added a note at the end of the subsection:
```
.. note::

    The ``tag`` identifier can be any alphanumeric string, it is simply a label used to refer to a previous vertex along the query path for defining a relationship.
```

Next, we look for all ``Int`` nodes that are an output of the ``CalcJobNode``'s found in the first step, using the ``with_incoming`` relationship argument.

> Since the ``CalcJobNode`` *created* the ``Int`` node, it is an *incoming* node from the ``Int`` node's perspective.

Author: I see, good to know, thanks!

In the context of our query, we are building a *path* consisting of *vertices* (i.e. the entities we query for) connected by *edges* defined by the relationships between them.

> A list of all the relationships you can use for your query, as well as the entities that they connect, can be found in the ``QueryBuilder`` docstring.

Reviewer: ```suggestion
The complete set of all possible relationships you can query for, as well as the entities that they connect to, is documented as part of the ``QueryBuilder`` class documentation.
```

> A list of all the relationships you can use for your query, as well as the entities that they connect, can be found in the ``QueryBuilder`` docstring.

Reviewer: If they are part of the reference API docs, they *are* in the docs, no?

> A list of all the relationships you can use for your query, as well as the entities that they connect, can be found in the ``QueryBuilder`` docstring.

Reviewer 2: Technically yes, but they are not very readable. The auto generated API docs are very difficult to navigate as they literally print everything. I think someone suggested another plugin that tries to ameliorate this by providing a TOC for each class to make it more navigatable. But until that is implemented I find that relying on the docstrings should be minimal. This table of all operators is not some thing that is only useful for fringe use cases, but is really important if you want to do any kind of meaningful querying. I therefore think it deserves a visible and well formatted spot in the normal documentation

> A list of all the relationships you can use for your query, as well as the entities that they connect, can be found in the ``QueryBuilder`` docstring.

Reviewer: Like you said, this is more an issue with the API docs. I think the API docs should be *complete*, but we are currently just dumping literally *everything* into the docs, whether it is considered public API or not. This is something that we should address, but probably *after* the revamp.

So, then let's make it its own sub-section for now and then potentially replace it at some point?

> A list of all the relationships you can use for your query, as well as the entities that they connect, can be found in the ``QueryBuilder`` docstring.

Author: Alright, I will move the tables to the "advanced query" subsection in the topics for now. We can still settle on a final location after the revamp! :)

.. _how-to:data:find:projections:

Projections

> -----------

Reviewer: I think it would be worthwhile to super briefly motivate "projections" again in one sentence at the start of this sub-section.

> When no *projection* is specified, the default behavior of the ``QueryBuilder`` is to project the entities corresponding to the final vertex of the query path.

Reviewer 2: This sentence is a bit complicated and potentially confusing as it mixes two parts of the behavior.
By default, if no explicit projections are specified, the builder will return:
1. entities that were appended last
2. actual instances of the entities

You first sentence refers to the first, but the "however" in your following sentence applies to the second point.
I think the solution is to move the idea of specific property projections to a later point and first explain how to project an entity that was not appended last. You show `project='*'` which then leads nicely into the next section that `*` projects the entire instance, but you can also project just certain properties. Note that I don't think we should call it `attributes` because `label`, `pk`, etc. can also be projected but are not attbributes.

However, in many cases we are not interested in the entities themselves, but rather their PK, UUID, *attributes* or some other piece of information stored by the entity.
This can be achieved using the ``project`` keyword argument:

.. code-block:: python

    qb = QueryBuilder()
    qb.append(CalcJobNode, tag='calcjob')
    qb.append(Int, with_incoming='calcjob', project='id')

In the above example, executing the query returns all PK's of the ``Int`` nodes which are outputs of all ``CalcJobNode``'s in the database.
However, you can project more than one piece of information for one vertex by using a list:

.. code-block:: python

    qb = QueryBuilder()
    qb.append(CalcJobNode, tag='calcjob')
    qb.append(Int, with_incoming='calcjob', project=['id', 'attributes.value'])

Moreover, you can project information for multiple vertices along the query path:

.. code-block:: python

    qb = QueryBuilder()
    qb.append(CalcJobNode, tag='calcjob', project='uuid)
    qb.append(Int, with_incoming='calcjob', project=['id', 'attributes.value'])

Asking only for the properties that you are interested in can result in much faster queries.

> If you want the Aiida-ORM instance, add ``'*'`` to your list of projections:

Reviewer: What exactly does this refer to in this context? The node itself?

Author: Yes. I'll try and make this more clear when rewriting the subsection on projections with regard to @sphuber's comment below.

.. code-block:: python

    qb = QueryBuilder()
    qb.append(CalcJobNode, tag='calcjob', project='uuid')
    qb.append(Int, with_incoming='calcjob', project=['id', '*', 'attributes.value'])

All projections must start with one of the *columns* of the entities in the database.
Examples of columns we have encountered so far are ``id``, ``uuid`` and ``attributes``.
If the column is a dictionary, you can expand the dictionary values using a dot notation, as we have done in the previous example to obtain the ``attributes.value``.
This can be used to project the values of nested dictionaries as well.

.. note::

>     Be aware that, for consistency, ``QueryBuilder.all()`` / ``iterall()`` always returns a list of lists, even if you project on one entity.

Reviewer: ```suggestion
    Be aware that for consistency, ``QueryBuilder.all()`` / ``iterall()`` always returns a list of lists, even if you project on one entity.
```

>     A convenient boolean keyword argument for the ``all()`` method is ``flat``, which returns the query as a flat list if set to ``True``.

Reviewer: ```suggestion
    Use ``all(flat=True)`` to return the query result as a flat list.
```

>     A convenient boolean keyword argument for the ``all()`` method is ``flat``, which returns the query as a flat list if set to ``True``.

Reviewer 2: should we point out that this really often only makes sense if you are projecting a single property of a single entity?

Author: I've adjusted the first sentence to highlight this some more, and added "in this case" to the second:

```
.. note::

    Be aware that, for consistency, ``QueryBuilder.all()`` / ``iterall()`` always returns a list of lists, even if you only project one property of a single entity.
    A convenient boolean keyword argument for the ``all()`` method in this case is ``flat``, which returns the query as a flat list if set to ``True``.
```

> As mentioned in the introduction of this how-to, these sections only explain the basic functionality of the ``QueryBuilder``.
> For more advanced queries, please see :ref:`the corresponding topics section<topics:database:advancedquery>`.

Reviewer: ```suggestion
As mentioned in the beginning, this section provides only a brief introduction to the ``QueryBuilder``'s basic functionality.
To learn about more advanced queries, please see :ref:`the corresponding topics section<topics:database:advancedquery>`.
```

However, you can also *project* other entities in the path by adding ``project='*'`` to the corresponding ``append()`` call:

.. code-block:: python

    qb = QueryBuilder()
    qb.append(CalcJobNode, tag='calcjob', project='*')
    qb.append(Int, with_incoming='calcjob')

This will return all  :class:`~aiida.orm.nodes.process.calculation.calcjob.CalcJobNode`'s that have an ``Int`` output node.

However, in many cases we are not interested in the entities themselves, but rather their PK, UUID, *attributes* or some other piece of information stored by the entity.
This can be achieved by providing the corresponding *column* to the ``project`` keyword argument:

.. code-block:: python

    qb = QueryBuilder()
    qb.append(CalcJobNode, tag='calcjob')
    qb.append(Int, with_incoming='calcjob', project='id')

In the above example, executing the query returns all *PK's* of the ``Int`` nodes which are outputs of all  :class:`~aiida.orm.nodes.process.calculation.calcjob.CalcJobNode`'s in the database.
Moreover, you can project more than one piece of information for one vertex by providing a list:

.. code-block:: python

    qb = QueryBuilder()
    qb.append(CalcJobNode, tag='calcjob')

>     qb.append(Int, with_incoming='calcjob', project=['id', '*', 'attributes.value'])

Author: Fair point! The idea was to introduce `project='*'` to another query to show that you can use `'*'` like any other projection. I've moved this concept to the next query example.

.. _how-to:data:find:

Finding and querying for data
=============================

Once you have successfully completed a series of workflows for your project, or have imported a dataset you are interested in, you want to quickly find the data that is relevant for your analysis.

> To make it easier to find or *query* for the data you need, AiiDA provides the ``QueryBuilder`` class.

Author: How about replacing the sentence by this one:

> The data in an AiiDA database is stored as a graph of connected entities, which can be easily *queried* with the `QueryBuilder` class.

> To make it easier to find or *query* for the data you need, AiiDA provides the ``QueryBuilder`` class.

Reviewer: Yes, that sounds better.

^^^^^^^^^^^^^^^^^^

Once you have *appended* the entity you want to query for to the ``QueryBuilder``, the next question is how to get the results.
There are several ways to obtain data from a query::

    qb = QueryBuilder()                 # Instantiating instance
    qb.append(CalcJobNode)              # Setting first vertice of path

    first_row = qb.first()              # Returns a list (!)
                                        # of the results of the first row

    all_results_d = qb.dict()           # Returns all results as
                                        # a list of dictionaries

    all_results_l = qb.all()            # Returns a list of lists

In case you are working with a large dataset, you can also return your query as a generator::

    all_res_d_gen = qb.iterdict()       # Return a generator of dictionaries
                                        # of all results
    all_res_l_gen = qb.iterall()        # Returns a generator of lists

This will retrieve the data in batches, and you can start working with the data before the query has completely finished.
Be aware that when using generators, you should never commit (store) anything during the iteration.

> While the query is still running it can be compromised by new data in the database.

Author: I don't think the data is compromised. The way I understand the sentence is that the query can be compromised when storing new data during the query.

I'll add your suggested sentence as a warning for now, but leave this comment as unresolved.

> While the query is still running it can be compromised by new data in the database.

Reviewer: @sphuber Could you clarify on what exactly is at risk of being compromised here?

> While the query is still running it can be compromised by new data in the database.

Reviewer 2: I don't think that existing data can really be "compromised" or corrupted in so much as that updating a collection when iterating over it, is in general not advisable (note that comprised in the text refers to the query and not the data in the database). This also holds for Python itself, since the operation, fundamentally, is not well defined and *could* have unexpected behavior. Python even raises for this kind of behavior:
```
In [3]: d = {'a': 1, 'b': 2, 'c': 3}                                                                                                                                                                             

In [4]: for key in d: 
   ...:     if key == 'a': 
   ...:         d.pop('b') 
   ...:                                                                                                                                                                                                          
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-4-ce771193fc04> in <module>
----> 1 for key in d:
      2     if key == 'a':
      3         d.pop('b')
      4 

RuntimeError: dictionary changed size during iteration
```

To conclude: we could say that users should not create new entities when iterating over the relevant collection as the query might not make sense at some point. However, I think the chances of anyone doing this by accident is minimal, and at worst the query will crash. Don't think there will be much worse consequences but can not be certain

> While the query is still running it can be compromised by new data in the database.

Author: So, perhaps it's not necessary to add a warning in the how-to, but rather move this discussion to the topics section?

> While the query is still running it can be compromised by new data in the database.

Reviewer: Yeah, I don't think this is enough of a problem to confuse users here.

> While the query is still running it can be compromised by new data in the database.

Reviewer: This could also be mentioned as part of the doc-string.

    |            |             |                                          | than the specified value         |
    |            |             |                                          | **(currently implemented only    |
    |            |             |                                          | in the SQLA backend)**           |
    +------------+-------------+------------------------------------------+----------------------------------+
    | contains   |    lists    | | 'attributes.mykey': {'contains':       | Check that a list (typically     |
    |            |             |   ['a','b']}                             | stored in the attributes or in   |
    |            |             |                                          | the extras) contains some        |
    |            |             |                                          | specific elements or values      |
    |            |             |                                          | **(currently implemented only    |
    |            |             |                                          | in the SQLA backend)**           |

>     +------------+-------------+------------------------------------------+----------------------------------+

Reviewer 2: For my group meeting presentation, I reformatted these two tables into one and correctly used inline code tags for the operators and examples and fixed the formatting/spacing. I am pasting it here so you can include it:
```

This is a list of all implemented operators:

+--------------+-------------+-------------------------------------------------------+------------------------------------------------------------------------------+
|**Operator**  |**Datatype** |  **Example**                                          | Explanation                                                                  |
+==============+=============+=======================================================+==============================================================================+
|   ``==``     |      all    | ``'id': {'==': 123}``                                 | Filter for equality                                                          |
+--------------+-------------+-------------------------------------------------------+------------------------------------------------------------------------------+
|   ``in``     |      all    | ``'name': {'in': ['foo', 'bar']}``                    | Filter for values that are in the given list.                                |
+--------------+-------------+-------------------------------------------------------+------------------------------------------------------------------------------+
| ``>,<,<=,>=``| float,      | ``'ctime': {'<': datetime(2016, 03, 03)}``            | Filter for values that are greater or smaller than a certain value           |
|              | integer,    |                                                       |                                                                              |
|              | date        |                                                       |                                                                              |
+--------------+-------------+-------------------------------------------------------+------------------------------------------------------------------------------+
| ``like``     | string      | ``'name': {'like': 'label%'}``                        | Filter for matching substrings where ``%`` and ``_`` are wildcards.          |
|              |             |                                                       | To match a literal ``%`` or ``_`` escape it by prefixing it with ``\\``.     |
|              |             |                                                       |                                                                              |
|              |             |                                                       |                                                                              |
+--------------+-------------+-------------------------------------------------------+------------------------------------------------------------------------------+
| ``ilike``    | string      | ``'name': {'ilike': 'lAbEl%'}``                       | Case insensitive version of ``like``.                                        |
+--------------+-------------+-------------------------------------------------------+------------------------------------------------------------------------------+
| ``or``       | list of     | ``'id': {'or': [{'<': 12}, {'==': 199}]}``            | A list of expressions where at least one should be matched.                  |
|              | expressions |                                                       |                                                                              |
+--------------+-------------+-------------------------------------------------------+------------------------------------------------------------------------------+
| ``and``      | list of     | ``'id': {'and': [{'<': 12}, {'>': 1}]}``              | A list of expressions where all should be matched.                           |
|              | expressions |                                                       |                                                                              |
+--------------+-------------+-------------------------------------------------------+------------------------------------------------------------------------------+
| ``has_key``  | dict        | ``'attributes': {'has_key': 'some_key'}``             | Filter for dictionaries that contain a certain key.                          |
+--------------+-------------+-------------------------------------------------------+------------------------------------------------------------------------------+
| ``of_type``  |    any      | ``'attributes.some_key': {'of_type': 'bool'}``        | Filter for values of a certain type.                                         |
+--------------+-------------+-------------------------------------------------------+------------------------------------------------------------------------------+
| ``of_length``|    lists    | ``'attributes.some_list': {'of_length': 4}``          | Filter for lists of a certain length.                                        |
+--------------+-------------+-------------------------------------------------------+------------------------------------------------------------------------------+
| ``shorter``  |    lists    | ``'attributes.some_list': {'shorter': 4}``            | Filter for lists that are shorter than a certain length.                     |
+--------------+-------------+-------------------------------------------------------+------------------------------------------------------------------------------+
| ``longer``   |    lists    | ``'attributes.some_list': {'longer': 4}``             | Filter for lists that are longer than a certain length.                      |
+--------------+-------------+-------------------------------------------------------+------------------------------------------------------------------------------+
| ``contains`` |    lists    | ``'attributes.some_key': {'contains': ['a', 'b']}``   | Filter for lists that should contain certain values.                         |
+--------------+-------------+-------------------------------------------------------+------------------------------------------------------------------------------+
```

Author: Great, thanks!

    qb = QueryBuilder()
    qb.append(CalcJobNode, tag='calcjob')
    qb.append(Int, with_outgoing='calcjob', edge_filters={'label': 'x'})

By using the ``edge_filters`` keyword argument, we can query for only the inputs that have the label ``x``.
Similarly, we can project information of the edge using the ``edge_project`` keyword argument::

    qb = QueryBuilder()
    qb.append(CalcJobNode, tag='calcjob')
    qb.append(Int, with_outgoing='calcjob', edge_project='label')

In the example above, we are querying for the edge labels of the incoming ``Int`` nodes of all ``CalcJobNode``'s.
Note that you can use the :ref:`operators from the table above <how-to:data:find:appender:operators>` to filter edges as well.
Say we want to find all input ``Int`` nodes that do are **not** connected to the ``CalcJobNode``'s via an edge with label ``x``::

    qb = QueryBuilder()
    qb.append(CalcJobNode, tag='calcjob')
    qb.append(Int, with_outgoing='calcjob', edge_filters={'label': {'!==': 'x'}})

.. how-to:data:find:appender:ordering:

Ordering and limiting results
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> You can also order the results of your query by the properties of the entity.

Reviewer: ```suggestion
You can order the results of your query by the properties of the entity.
```

Note that in this case you have to specify the order operation with a dictionary, where the ``order`` key details the way you want to order the query results and the ``cast`` key informs the ``QueryBuilder`` of the attribute type.
Below is a list of the available cast types and their aliases:

+-------------------+-----------+---------------------+
| **Python type**   | **Alias** | **SQLAlchemy type** |
+===================+===========+=====================+
| int               | i         | Integer             |
+-------------------+-----------+---------------------+
| float             | f         | Float               |
+-------------------+-----------+---------------------+
| bool              | b         | Boolean             |
+-------------------+-----------+---------------------+
| str               | t         | String              |
+-------------------+-----------+---------------------+
| dict              | j         | JSONB               |
+-------------------+-----------+---------------------+
| datetime.datetime | d         | DateTime            |

> +-------------------+-----------+---------------------+

Reviewer: I think this should be in the reference section.

You can also order using multiple properties by providing a list of dictionaries that each specify one sorting operation::

    qb = QueryBuilder()
    qb.append(CalcJobNode, tag='calcjob')
    qb.append(Int, with_incoming='calcjob')
    qb.order_by({Int: [{'attributes.value': {'order': 'asc', 'cast': 'f'}}, {'ctime': 'desc'}]})
    qb.all()

Here the ``Int`` nodes will first be sorted by their value in ascending order.
Nodes for which the value is equal are subsequently sorted by their modification time in descending order.

Finally, you can also limit the number of query results returned with the ``limit`` method.

> Suppose you only want the three first results from our query::

Reviewer: ```suggestion
Suppose you only want the first three results from our query::
```

        That other node can be specified by an integer or the class or type.
        The following examples are all valid joining instructions, assuming there is a structure defined at index 2 of the path with tag "struc1"::

            edge_specification = queryhelp['path'][3]
            edge_specification['with_incoming'] = 2
            edge_specification['with_incoming'] = StructureData
            edge_specification['with_incoming'] = 'struc1'
            edge_specification['with_outgoing']  = 2
            edge_specification['with_outgoing']  = StructureData
            edge_specification['with_outgoing']  = 'struc1'

    *   queryhelp_item['direction'] = integer

        If any of the above specs ("with_outgoing", "with_incoming") were not specified, the key "direction" is looked for.
        Directions are defined as distances in the tree.
        1 is defined as one step down the tree along a link.
        This means that 1 joins the node specified in this dictionary to the node specified on list-item before **as an output**.
        Direction defaults to 1, which is why, if nothing is specified, this node is joined to the previous one as an output by default.

>         A minus sign reverse the direction of the link.

Reviewer: ```suggestion
        A negative number reverse the direction of the link.
```

        The absolute value of the direction defines the table to join to with respect to your own position in the list.
        An absolute value of 1 joins one table above, a value of 2 to the table defined 2 indices above.

>         The two following queryhelps yield the same  query::

Reviewer: ```suggestion
        The two following queryhelps yield the same query::
```

            from aiida.orm import TrajectoryData
            from aiida_quantumespresso.calculations.pw import PwCalculation
            from aiida.orm import Dict
            qh1 = {

>                 'path':[

Reviewer: We should follow the convetion of one space after the `:`:
```suggestion
                'path': [
```
Same goes for all the other occurrences.

                    {
                        'cls':Dict,
                        'with_outgoing':PwCalculation
                    }
                ]
            }

            # Shorter version:

            qh3 = {
                'path':[
                    Dict,
                    PwCalculation,
                    TrajectoryData,
                ]
            }

*   what to ``project``: Determing which columns the query will return::

        queryhelp = {
            'path':[PwCalculation],
            'project':{
                PwCalculation:['user_id', 'id'],
            }
        }

    If you are using JSONB columns, you can also project a value stored inside the json::

        queryhelp = {
            'path':[
                PwCalculation,
                StructureData,
            ],
            'project':{
                PwCalculation:['state', 'id'],
            }
        }

    Returns the state and the id of all instances of ``PwCalculation`` where a structures is linked as output of a relax-calculation.
    The strings that you pass have to be name of the columns.
    If you pass an asterisk ('*'), the query will return the instance of the AiidaClass.

> 

Reviewer: ```suggestion
```

*   the ``filters``:

>     What if you want not every structure, but only the ones that were added after a certain time (say last 4 days) and have an id higher than 50::

Reviewer: ```suggestion
    Filters enable you to further specify the query.
    This is an example for a query for structures that were added after a certain time (say last 4 days) and have an id larger than 50::
```

        from aiida.common import timezone
        from datetime import timedelta

        queryhelp = {
            'path':[
                {'cls':PwCalculation}, # PwCalculation with structure as output
                {'cls':StructureData}
            ],
            'filters':{
                StructureData:{
                    'ctime':{'>':  timezone.now() - timedelta(days=4)},
                    'id':{'>': 50}
                }
            }
        }

If you want to include filters and projections on links between nodes, you will have to add these to filters and projections in the queryhelp.

> Let's take an example that we had and add a few filters on the link::

Reviewer: ```suggestion
Let's take an example from before and add a few filters on the link::
```

    queryhelp = {
        'path':[
            {'cls':PwCalculation, 'tag':'relax'}, # PwCalculation with structure as output
            {'cls':StructureData, 'tag':'structure'}
        ],
        'filters':{
            'structure':{
                'id':{'>': 50}
            },
            'relax--structure':{
                'label':{'like':'output_%'},
            }
        },
        'project':{
            'relax--structure':['label'],
            'structure':['label'],
            'relax':['label', 'uuid'],
        }
    }

Notice that the tag for the link, by default, is the tag of the two connecting nodes delimited by two dashes '--' and the order DOES matter.

> If you dislike that way to tag the link, you can choose the tag for the edge in the path when definining the entity to join using ``edge_tag``::

Reviewer: I'm not sure that "disliking" a proposed best practice provides sufficient motivation.
```suggestion
Alternatively, you can choose the tag for the edge in the path when defining the entity to join using ``edge_tag``::
```

    queryhelp = {
        'path':[
            {'cls':PwCalculation, 'tag':'relax'},         # Relaxation with structure as output
            {
                'cls':StructureData,
                'tag':'structure',
                'edge_tag':'ThisIsMyLinkTag'     # Definining the link tag
            }
        ],
        'filters':{
            'structure':{
                'id':{'>': 50}
            },
            'ThisIsMyLinkTag':{                  # Using this link tag
                'label':{'like':'output_%'},
            }
        },
        'project':{
            'ThisIsMyLinkTag':['label'],
            'structure':['label'],
            'relax':['label', 'uuid'],
        }
    }

> You can set a limit and an offset in the queryhelp::

Reviewer: ```suggestion
Limits and offset can be set directly like this::
```

.. _#3996: https://github.com/aiidateam/aiida-core/issues/3996
.. _#3997: https://github.com/aiidateam/aiida-core/issues/3997

> .. _#3998: https://github.com/aiidateam/

Author: Yes, this was accidental. I must have gotten a little overzealous while slimming down the section. ^^
